truffle/src/main/java/org/jruby/truffle/nodes/core/KernelNodes.java

```diff
@@ -945,14 +949,42 @@ public boolean isANil(RubyBasicObject self, Object nil) {
             return false;
         }
 
-        @TruffleBoundary
+        @Specialization(
+                limit = "getCacheLimit()",
+                guards = {"getMetaClass(frame, self) == cachedMetaClass", "module == cachedModule"},
+                assumptions = "cachedModule.getUnmodifiedAssumption()")
+        public boolean isACached(VirtualFrame frame,
+                                 Object self,
+                                 RubyModule module,
+                                 @Cached("getMetaClass(frame, self)") RubyClass cachedMetaClass,
+                                 @Cached("module") RubyModule cachedModule,
+                                 @Cached("IsA(cachedMetaClass, cachedModule)") boolean result) {
+            return result;
+        }
+
         @Specialization
-        public boolean isA(Object self, RubyModule rubyClass) {
+        public boolean isAUncached(VirtualFrame frame, Object self, RubyModule module) {
+            return IsA(getMetaClass(frame, self), module);
+        }
+
+        @Specialization(guards = "!isRubyModule(module)")
+        public boolean isATypError(VirtualFrame frame, Object self, Object module) {
             CompilerDirectives.transferToInterpreter();
-            // TODO(CS): fast path
-            return ModuleOperations.assignableTo(getContext().getCoreLibrary().getMetaClass(self), rubyClass);
+            throw new RaiseException(getContext().getCoreLibrary().typeError("class or module required", this));
         }
 
+        @TruffleBoundary
+        protected boolean IsA(RubyClass metaClass, RubyModule module) {
+            return ModuleOperations.assignableTo(metaClass, module);
+        }
+
+        protected RubyClass getMetaClass(VirtualFrame frame, Object object) {
+            return metaClassNode.executeMetaClass(frame, object);
+        }
+
+        protected int getCacheLimit() {
+            return LookupConstantNode.getCacheLimit();
```

Comment:
Maybe it is better to reference the original property (= inline the code). I think these getCacheLimit() should probably be all protected.